sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,924 @@
+                cli_logger.error(f"""Create VM instance for {name}
+                    failed with status code {str(e.code)}
+                    .\nFailed instance prototype:\n""")
+                pprint(instance_prototype)
+            raise e
+
+        logger.info(f"VM instance {name} created successfully")
+        return resp.result
+
+    def _create_floating_ip(self, base_config):
+        """returns unbound floating IP address.
+        Creates a new ip if none were found in the config file.
+        Args:
+            base_config(dict): specific node relevant data.
+            node type segment of the cluster's config file,
+            e.g. ray_head_default.
+        """
+
+        if base_config.get("head_ip"):
+            # pylint: disable=line-too-long
+            for ip in self.ibm_vpc_client.list_floating_ips().get_result()["floating_ips"]:
+                if ip["address"] == base_config["head_ip"]:
+                    return ip
+
+        floating_ip_name = f"{RAY_RECYCLABLE}-{uuid4().hex[:4]}"
+        # create a new floating ip
+        logger.debug(f"Creating floating IP {floating_ip_name}")
+        floating_ip_prototype = {}
+        floating_ip_prototype["name"] = floating_ip_name
+        floating_ip_prototype["zone"] = {"name": self.zone}
+        floating_ip_prototype["resource_group"] = {
+            "id": self.resource_group_id
+        }
+        response = self.ibm_vpc_client.create_floating_ip(
+            floating_ip_prototype)
+        floating_ip_data = response.result
+
+        return floating_ip_data
+
+    def _attach_floating_ip(self, instance, fip_data):
+        """
+        attach a floating ip to the network interface of an instance
+        Args:
+            instance(dict): extensive data of a node.
+            fip_data(dict): floating ip data.
+        """
+
+        fip = fip_data["address"]
+        fip_id = fip_data["id"]
+
+        logger.debug(f"""Attaching floating IP {fip} to
+            VM instance {instance["id"]}""")
+
+        # check if floating ip is not attached yet
+        inst_p_nic = instance["primary_network_interface"]
+
+        if inst_p_nic["primary_ip"] and inst_p_nic["id"] == fip_id:
+            # floating ip already attached. do nothing
+            logger.debug(f"Floating IP {fip} already attached to eth0")
+        else:
+            # attach floating ip
+            self.ibm_vpc_client.add_instance_network_interface_floating_ip(
+                instance["id"], instance["network_interfaces"][0]["id"], fip_id
+            )
+
+    def _stopped_nodes(self, tags):
+        """
+        returns stopped nodes of type specified in tags.
+        TAG_RAY_NODE_KIND is a mandatory field.
+        Args:
+            tags(dict): set of conditions nodes will be filtered by.
+        """
+        # pylint: disable=C0206 W0622
+        filter = {
+            TAG_RAY_CLUSTER_NAME: self.cluster_name,
+            TAG_RAY_NODE_KIND: tags[TAG_RAY_NODE_KIND],
+        }
```

Comment:
i remember asking the same question a year ago. will get rid of `filter` and use the provided `tags` instead.

Reply:
After some debugging, there's most certainly a need for it. i've added the following documentation to the function: 
```
  # the following subset 'filter' is a common denominator for all
  # nodes belonging to the cluster.
  # can't filter using tags in its entirety, since worker nodes might
  # be created by ray with the most basic tags.
  # the entire set of tags of these stopped nodes will be updated on the
  # calling function: create_node().
``` 
Will resolve conversation upon your reply.

Reply:
Hmm I think I'm confused. Isn't the behavior of this function to return stopped nodes filtered by tags and not all stopped nodes?

Reply:
The purpose of the function is to return all stopped nodes. 
Ray calls `create_node` with tags to set to the newly created / restarted nodes. 
The only key we should take into account when restarting a stopped node is kind of node, i.e. whether Ray would like to restart a head or worker node. Even the cluster name isn't really required as we already keep track of it, using the one to one cluster to VPC relationship, where the VPC is tagged with the cluster name.

I'll alter the documentation. I see now that it's misleading.